corehq/apps/case_importer/do_import.py

```diff
@@ -25,248 +25,262 @@
-        'errors': errors.as_dict(),
-        'num_chunks': num_chunks,
-    }
+                track_load()
+                if parent_case:
+                    extras['index'] = {
+                        parent_ref: (parent_type, parent_case.case_id)
+                    }
+
+            case_name = fields_to_update.pop('name', None)
+
+            if BULK_UPLOAD_DATE_OPENED.enabled(domain):
+                date_opened = fields_to_update.pop(CASE_TAG_DATE_OPENED, None)
+                if date_opened:
+                    extras['date_opened'] = date_opened
+
+            if not case:
+                id = uuid.uuid4().hex
+
+                if config.search_field == 'external_id':
+                    extras['external_id'] = search_id
+                elif external_id:
+                    extras['external_id'] = external_id
+
+                try:
+                    caseblock = CaseBlock(
+                        create=True,
+                        case_id=id,
+                        owner_id=owner_id,
+                        user_id=user_id,
+                        case_type=config.case_type,
+                        case_name=case_name or '',
+                        update=fields_to_update,
+                        **extras
+                    )
+                    caseblocks.append(RowAndCase(i, caseblock))
+                    self.created_count += 1
+                    if external_id:
+                        ids_seen.add(external_id)
+                except CaseBlockError:
+                    self.errors.add(ImportErrors.CaseGeneration, i + 1)
+            else:
+                if external_id:
+                    extras['external_id'] = external_id
+                if uploaded_owner_id:
+                    extras['owner_id'] = owner_id
+                if to_close == 'yes':
+                    extras['close'] = True
+                if case_name is not None:
+                    extras['case_name'] = case_name
+
+                try:
+                    caseblock = CaseBlock(
+                        create=False,
+                        case_id=case.case_id,
+                        update=fields_to_update,
+                        **extras
+                    )
+                    caseblocks.append(RowAndCase(i, caseblock))
+                    self.match_count += 1
+                except CaseBlockError:
+                    self.errors.add(ImportErrors.CaseGeneration, i + 1)
+
+            # check if we've reached a reasonable chunksize
+            # and if so submit
+            if len(caseblocks) >= chunksize:
+                _submit_caseblocks(domain, config.case_type, caseblocks)
+                self.num_chunks += 1
+                caseblocks = []
+
+        # final purge of anything left in the queue
+        if _submit_caseblocks(domain, config.case_type, caseblocks):
+            self.match_count -= 1
+        self.num_chunks += 1
+
+    @property
+    def outcome(self):
```

Comment:
Do you mean like whether or not _any_ rows completed successfully?  Yes, I do think that needs to be addressed, as the way it's handled now is often wrong.  This would be the right place for that to live, I think.

Reply:
That was a bit cryptic, sorry.
The way this `_Importer` class is set up, the `_Importer(...).outcome()` will be meaningless unless you've already called `_Importer(...).do_import()`. I was suggesting that you explicitly say that somehow, or just make the call to `do_import` from this `outcome()` method. That way this class only has a single publicly accessible method and everything else is hidden tidily away.

Reply:
I do like the idea of fewer public methods, but the name "outcome()" doesn't make me think that calling it will also do an import (although that could just be me). For fewer methods, tidying everything else away, I'd lean towards renaming "outcome()" to "_outcome()" and having "do_import()" return it.

Reply:
This was changed around a bit in a subsequent commit - here's how the main function works now:

https://github.com/dimagi/commcare-hq/blob/3f2439c95be28188e417ccb95e8a5888a6f8bfa6/corehq/apps/case_importer/do_import.py#L38-L41

It still has essentially the same pitfalls as y'all are describing, though.  Let me explain my reasoning a little, I think we're all on roughly the same page.

This whole file can be summed up by that simple interface above - a function which accepts a handful of arguments and returns a simple JSON representation of the result.  However, far too much happens to fit that in a single function (hence this PR).

Given the amount of shared state, I think hosting it on a class was the most straightforward way to break it up.  This presents problems, as we really only want that simple function interface, but classes aren't supposed to do work in the constructor, and if you have fewer than two public methods, you don't need a class.  What I settled on was using a private class coupled tightly to the `do_import` function, which knows how to drive it.  This gives the external interface I want, and makes it more acceptable to bend some rules.  I do like returning the result directly from `do_import` though, I'll make that change.

Does that sound like the best way to approach this situation?  Simple wrapper function that instantiates the class and calls the `run()` method, with everything else hidden inside the class?

Reply:
Another option is this:
```python
class do_import(object):
    def __init__(self, spreadsheet, config, domain, task=None, record_form_callback=None):
        ...
        for row in spreadsheet:
            self.import_row
        return self.results.to_json()
```
but that just feels so wrong.

Reply:
> but that just feels so wrong.

**\*shudder\*** ... yes

> if you have fewer than two public methods, you don't need a class.

:thinking: Because Uncle Bob says so, because convention, or because it has been shown that you can write the same code functionally and it will be simpler? (The last one does sound possible, but I can't see whether it would always be true.)

Reply:
>  I do like returning the result directly from do_import though, I'll make that change.

I like this option the best, and is kind of the pattern I've been following when doing single function -> class refactors. 

> if you have fewer than two public methods, you don't need a class

Says who? I think you can even have _no_ public methods and still have a useful, easy to read class that acts as a data wrapper.

Reply:
There's a pylint warning `too-few-public-methods`.  I agree with it in principle, with the exception of things like dataclasses.  I do think in this case the one master method approach yields the best result.  I could even see adopting some kind of naming convention around this to make it clearer that this is a class that acts as a single function, though I don't wanna walk down that road right now.